core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientWithDecoratorTest.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.linecorp.armeria.client.WebClient;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.logging.RequestLogProperty;
+import com.linecorp.armeria.internal.testing.AnticipatedException;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+class RetryingClientWithDecoratorTest {
+
+    @RegisterExtension
+    static final ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/hello", (ctx, req) -> HttpResponse.of(200));
+        }
+    };
+
+    @Test
+    void responseCauseIsSetWhenExceptionIsRaisedInDecorator() throws InterruptedException {
+        final AtomicInteger onExceptionCounter = new AtomicInteger();
+        final AtomicInteger logExceptionCounter = new AtomicInteger();
+        // Retry only 3 times.
+        final RetryRule retryRule = RetryRule.builder()
+                                             .onException((ctx, cause) -> {
```

Comment:
This question may be out of the scope of this PR. Instead, I hope to know Armeria's expected behavior/semantics.

Should/will this `onException` callback be invoked with `cause != null` when a network-level exception happen?

By "network-level exception", I mean one of below situations:

### 1. Connection socket timeout.
E.g. the client tries to connect to a private address `10.255.255.255`.
Thrown exception:
`StatusRuntimeException` -> `UnprocessedRequestException` -> `ConnectTimeoutException`
### 2. Request timeout.
E.g. Server side is too slow and fails to response within a certain period of time.
Thrown exception:
`StatusRuntimeException` -> `ResponseTimeoutException` 
### 3. Connection interrupted.
E.g. server crashes unexpectedly in the middle of transmitting a response.
Thrown exception:
`StatusRuntimeException` -> `UnprocessedRequestException` -> `ClosedSessionException`

### 4. DNS failure.
Client tries to resolve an invalid domain name.
Thrown exception:
`StatusRuntimeException` -> `UnprocessedRequestException` -> `UnknownHostException`

Reply:
> Should/will this onException callback be invoked with cause != null when a network-level exception happen?

Yes, I believe so. If the exception is raised before the request goes to the network layer, the cause will be set by this change. If the exception is raised in the network layer or after that, the cause will be set [here](https://github.com/line/armeria/blob/cae3b6da373101c0976e2d00f1f3cbb706d07e00/core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java#L331-L335) thus we can use it.

Reply:
> By "network-level exception", I mean one of below situations:

In addition to the situations you listed, I recommend you check some cases where `UnprocessedRequestException.of()` is used.